src/io/kqueue.rs

```diff
@@ -0,0 +1,566 @@
+
+        /// Clear the event after it is triggered.
+        const ONESHOT = c::EV_ONESHOT as _;
+
+        /// TODO
+        const CLEAR = c::EV_CLEAR as _;
+
+        /// TODO
+        const EOF = c::EV_EOF as _;
+
+        /// TODO
+        const ERROR = c::EV_ERROR as _;
+    }
+}
+
+bitflags::bitflags! {
+    /// The flags for a virtual node event.
+    pub struct VnodeEvents : u32 {
+        /// The file was deleted.
+        const DELETE = c::NOTE_DELETE;
+
+        /// The file was written to.
+        const WRITE = c::NOTE_WRITE;
+
+        /// The file was extended.
+        const EXTEND = c::NOTE_EXTEND;
+
+        /// The file had its attributes changed.
+        const ATTRIBUTES = c::NOTE_ATTRIB;
+
+        /// The file was renamed.
+        const RENAME = c::NOTE_RENAME;
+
+        /// Access to the file was revoked.
+        const REVOKE = c::NOTE_REVOKE;
+    }
+}
+
+#[cfg(feature = "process")]
+bitflags::bitflags! {
+    /// The flags for a process event.
+    pub struct ProcessEvents : u32 {
+        /// The process exited.
+        const EXIT = c::NOTE_EXIT;
+
+        /// The process forked itself.
+        const FORK = c::NOTE_FORK;
+
+        /// The process executed a new process.
+        const EXEC = c::NOTE_EXEC;
+
+        /// TODO
+        const TRACK = c::NOTE_TRACK;
+    }
+}
+
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+bitflags::bitflags! {
+    /// The flags for a user event.
+    pub struct UserFlags : u32 {
+        /// Ignore the user input flags.
+        const NOINPUT = c::NOTE_FFNOP;
+
+        /// Bitwise AND fflags.
+        const AND = c::NOTE_FFAND;
+
+        /// Bitwise OR fflags.
+        const OR = c::NOTE_FFOR;
+
+        /// Copy fflags.
+        const COPY = c::NOTE_FFCOPY;
+
+        /// Control mask for operations.
+        const CTRLMASK = c::NOTE_FFCTRLMASK;
+
+        /// User defined flags for masks.
+        const UDFMASK = c::NOTE_FFLAGSMASK;
+
+        /// Trigger the event.
+        const TRIGGER = c::NOTE_TRIGGER;
+    }
+}
+
+/// User-defined flags.
+///
+/// Only the lower 24 bits are used in this struct.
+#[repr(transparent)]
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct UserDefinedFlags(u32);
+
+#[cfg(any(
+    target_os = "macos",
+    target_os = "ios",
+    target_os = "watchos",
+    target_os = "dragonfly",
+    target_os = "freebsd",
+))]
+impl UserDefinedFlags {
+    /// Create a new `UserDefinedFlags` from a `u32`.
+    pub fn new(flags: u32) -> Self {
+        Self(flags & EVFILT_USER_FLAGS)
+    }
+
+    /// Get the underlying `u32`.
+    pub fn get(self) -> u32 {
+        self.0
+    }
+}
+
+/// A `kqueue`.
+pub struct Kqueue {
+    /// The underlying `kqueue` file descriptor.
+    fd: OwnedFd,
+}
```

Comment:
I think it's more idiomatic to have a `Kqueue` object you can call `new()` and `kevent()` on rather than a freestanding `kevent` function, if not also less of a footgun.